File: src/cpp-utils/crypto/kdf/Scrypt.cpp
#include "Scrypt.h"
#include "cpp-utils/crypto/kdf/SCryptParameters.h"
#include "cpp-utils/crypto/symmetric/EncryptionKey.h"
#include "cpp-utils/data/Data.h"
#include "cpp-utils/random/Random.h"
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vendor_cryptopp/scrypt.h>

using std::string;

namespace cpputils {

constexpr SCryptSettings SCrypt::ParanoidSettings;
constexpr SCryptSettings SCrypt::DefaultSettings;
constexpr SCryptSettings SCrypt::TestSettings;

namespace {
EncryptionKey _derive(size_t keySize, const std::string& password, const SCryptParameters& kdfParameters) {
    auto result = EncryptionKey::Null(keySize);

    const size_t status = CryptoPP::Scrypt().DeriveKey(
        static_cast<uint8_t*>(result.data()), result.binaryLength(),
        reinterpret_cast<const uint8_t*>(password.c_str()), password.size(),
        static_cast<const uint8_t*>(kdfParameters.salt().data()), kdfParameters.salt().size(),
        kdfParameters.n(), kdfParameters.r(), kdfParameters.p()
    );
    if (status != 1) {
        throw std::runtime_error("Error running scrypt key derivation. Error code: "+std::to_string(status));
    }

    return result;
}

SCryptParameters _createNewSCryptParameters(const SCryptSettings& settings) {
    return SCryptParameters(Random::PseudoRandom()->get(settings.SALT_LEN), settings.N, settings.r, settings.p);
}
}

SCrypt::SCrypt(const SCryptSettings& settingsForNewKeys)
        :_settingsForNewKeys(settingsForNewKeys) {
}

EncryptionKey SCrypt::deriveExistingKey(size_t keySize, const std::string& password, const Data& kdfParameters) {
    const SCryptParameters parameters = SCryptParameters::deserialize(kdfParameters);
    auto key = _derive(keySize, password, parameters);
    return key;
}

SCrypt::KeyResult SCrypt::deriveNewKey(size_t keySize, const std::string& password) {
    const SCryptParameters kdfParameters = _createNewSCryptParameters(_settingsForNewKeys);
    auto key = _derive(keySize, password, kdfParameters);
    return SCrypt::KeyResult {
        key,
        kdfParameters.serialize()
    };
}
}

File: crates/utils/src/crypto/kdf/scrypt/backends/scrypt.rs
use anyhow::Result;
// TODO Separate out InfallibleUnwrap from lockable and don't depend on lockable from this crate
use lockable::InfallibleUnwrap;

use crate::crypto::kdf::PasswordBasedKDF;
use crate::crypto::kdf::scrypt::params::ScryptParams;
use crate::crypto::kdf::scrypt::settings::ScryptSettings;
use crate::crypto::symmetric::EncryptionKey;

pub struct ScryptScrypt;

impl PasswordBasedKDF for ScryptScrypt {
    type Settings = ScryptSettings;
    type Parameters = ScryptParams;

    fn derive_key(key_size: usize, password: &str, kdf_parameters: &ScryptParams) -> EncryptionKey {
        let params = scrypt::Params::new(
            kdf_parameters.log_n(),
            kdf_parameters.r(),
            kdf_parameters.p(),
        )
        .expect("Invalid scrypt parameters");
        EncryptionKey::new(key_size, |key_data| {
            scrypt::scrypt(
                password.as_bytes(),
                kdf_parameters.salt(),
                &params,
                key_data,
            )
            .expect("Error in scrypt");
            Ok(())
        })
        .infallible_unwrap()
    }

    fn generate_parameters(settings: &ScryptSettings) -> Result<ScryptParams> {
        ScryptParams::generate(settings)
    }
}
